### PR TITLE
Clean up NonBacktracking doubly-linked changes

### PR DESCRIPTION
**Pull request by stephentoub**
I didn't get a chance to review https://github.com/dotnet/runtime/pull/68307 before it was merged.  This addresses my own feedback on it after the fact.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs [hunk 1]
```diff
@@ -90,7 +90,7 @@ internal Regex(string pattern, RegexOptions options, TimeSpan matchTimeout, Cult
             if ((options & RegexOptions.NonBacktracking) != 0)
             {
                 // If we're in non-backtracking mode, create the appropriate factory.
-                factory = new SymbolicRegexRunnerFactory(tree, options, matchTimeout, culture);
+                factory = new SymbolicRegexRunnerFactory(tree, options, matchTimeout);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs:93]: There was some IgnoreCase stuff in this implementation, and none of that is relevant anymore (the only place IgnoreCase exists in the RegexNode tree after parsing is with Backreferences, and SymbolicRegexRunnerFactory doesn't support backreferences).

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs [hunk 2]
```diff
@@ -40,16 +36,21 @@ internal SymbolicRegexNode<BDD> ConvertToSymbolicRegexNode(RegexNode root)
...
+            DoublyLinkedList<SymbolicRegexNode<BDD>> rootResult = new();
 
-            while (work.Count > 0)
+            // Create a stack to be processed in order to process iteratively rather than recursively, and push the root on.
+            Stack<(RegexNode Node, bool TryToMarkFixedLength, DoublyLinkedList<SymbolicRegexNode<BDD>> Result, DoublyLinkedList<SymbolicRegexNode<BDD>>[]? ChildResults)> stack = new();
+            stack.Push((root, true, rootResult, CreateChildResultArray(root.ChildCount())));
+
+            // Continue to iterate until the stack is empty, popping the next item on each iteration.
+            // Some popped items may be pushed back on as part of processing.
+            while (stack.TryPop(out (RegexNode Node, bool TryToMarkFixedLength, DoublyLinkedList<SymbolicRegexNode<BDD>> Result, DoublyLinkedList<SymbolicRegexNode<BDD>>[]? ChildResults) popped))
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs:48]: Every case except for one was popping after the peek.  I inverted it to always pop and just push the popped item back in the one case where that's needed.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs [hunk 3]
```diff
@@ -253,63 +231,70 @@ static string UnexpectedNodeType(RegexNode node)
...
                                         continue;
                                     }
 
-                                    or = SymbolicRegexNode<BDD>.OrderedOr(_builder, elem, or);
-
-                                    if (elem.IsAnyStar)
-                                        or = elem; // .* is the absorbing element
+                                    or = elem.IsAnyStar ?
+                                        elem : // .* is the absorbing element
+                                        SymbolicRegexNode<BDD>.OrderedOr(_builder, elem, or);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs:251]: I assume OrderedOr doesn't have side-effects, and so in the elem.IsAnyStar we can just skip the OrderedOr call.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs [hunk 4]
```diff
@@ -424,56 +409,56 @@ internal SymbolicRegexNode<TNewSet> Transform<TNewSet>(SymbolicRegexNode<TSet> s
...
-                        for (int i = 0; i < sr_elems.Count; i++)
+                        List<SymbolicRegexNode<TSet>> concatElements = node.ToList();
+                        SymbolicRegexNode<TNewSet>[] reverseTransformed = new SymbolicRegexNode<TNewSet>[concatElements.Count];
+                        for (int i = 0; i < reverseTransformed.Length; i++)
                         {
-                            sr_elems_trasformed[i] = Transform(sr_elems[i], builder, setTransformer);
+                            reverseTransformed[i] = Transform(concatElements[^(i + 1)], builder, setTransformer);
                         }
-                        return builder.CreateConcat(sr_elems_trasformed, false);
+                        return builder.CreateConcatAlreadyReversed(reverseTransformed, tryCreateFixedLengthMarker: false);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexBuilder.cs:451]: I reversed the order of filling the array here, such that we could then get rid of the CreateConcat overload which itself iterates through the array in reverse order and delegates to this one.

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs [hunk 5]
```diff
@@ -126,84 +116,71 @@ internal SymbolicRegexNode<BDD> ConvertToSymbolicRegexNode(RegexNode root)
...
-                                result.InsertAtEnd(_builder.CreateLoop(_builder.CreateSingleton(setBdd), node.Kind == RegexNodeKind.Setlazy, node.M, node.N));
+                                BDD setBdd = CreateBDDFromSetString(set);
+                                result.AddLast(_builder.CreateLoop(_builder.CreateSingleton(setBdd), node.Kind == RegexNodeKind.Setlazy, node.M, node.N));
                                 break;
                             }
 
                         case RegexNodeKind.Empty:
                         case RegexNodeKind.UpdateBumpalong: // UpdateBumpalong is a directive relevant only to backtracking and can be ignored just like Empty
-                            _ = work.Pop();
-                            result.InsertAtEnd(_builder.Epsilon);
```
Comment (stephentoub) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs:156]: I'm assuming there's no need to actually push epsilon into the list.

**Review by olsaarik — APPROVED** [review]
Looks good. I like that you managed to get rid of the reverse concat overload.

**Comment by stephentoub** [comment]
Thanks for reviewing.

**Pull request merged**